Digital Credential API needs a CredentialRequestCoordinator and clients skeletons

marcoscaceres opened the pull request:
#### 01e05cb09f9e7ce8b24e917384f9b0c750d0b4ed
<pre>
Digital Credential API needs a CredentialRequestCoordinator and clients skeletons
<a href="https://bugs.webkit.org/show_bug.cgi?id=270778">https://bugs.webkit.org/show_bug.cgi?id=270778</a>
<a href="https://rdar.apple.com/124874018">rdar://124874018</a>

Reviewed by Abrar Rahman Protyasha.

Implements CredentialRequestCoordinator and clients skeletons.

* LayoutTests/http/wpt/identity/identitycredentialscontainer-get-basics.https.html:
* LayoutTests/http/wpt/identity/identitycredentialscontainer-get-hidden.https.html:
* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/allow-attribute.https.html:
* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/enabled-on-self-origin-by-permissions-policy.https.sub.html:
* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/support/iframe.html:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp: Added.
(WebCore::CredentialRequestCoordinator::CredentialRequestCoordinator):
(WebCore::CredentialRequestCoordinator::discoverFromExternalSource):
* Source/WebCore/Modules/identity/CredentialRequestCoordinator.h: Copied from Source/WebCore/Modules/identity/IdentityCredentialsContainer.h.
* Source/WebCore/Modules/identity/CredentialRequestCoordinatorClient.h: Copied from Source/WebCore/Modules/identity/IdentityCredentialsContainer.h.
* Source/WebCore/Modules/identity/IdentityCredentialProtocol.h:
* Source/WebCore/Modules/identity/IdentityCredentialProtocol.idl:
* Source/WebCore/Modules/identity/IdentityCredentialsContainer.cpp:
(WebCore::IdentityCredentialsContainer::get):
* Source/WebCore/Modules/identity/IdentityCredentialsContainer.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Page.cpp:
* Source/WebCore/page/Page.h:
(WebCore::Page::credentialRequestCoordinator):
* Source/WebCore/page/PageConfiguration.cpp:
* Source/WebCore/page/PageConfiguration.h:

Canonical link: <a href="https://commits.webkit.org/281646@main">https://commits.webkit.org/281646@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ed222c1d4b7c4388401b0ca380fd58accc83b3b9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60483 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39839 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13050 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64406 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11020 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62613 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47515 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11251 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48944 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7657 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62514 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37107 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52396 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29777 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33804 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9620 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9935 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55686 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9916 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66137 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4421 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9745 "Found 60 new test failures: accessibility/mac/replace-text-with-empty-range.html accessibility/mac/replace-text-with-range-value-change-notification.html accessibility/mac/replace-text-with-range.html animations/cross-fade-border-image-source.html css3/blending/background-blend-mode-svg.html css3/color/backgrounds-and-borders.html css3/color/box-shadows.html css3/color/canvas.html css3/color/composited-solid-backgrounds.html css3/color/gradients.html ... (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56307 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4440 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52367 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56474 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13472 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3670 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35646 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36728 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37819 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36473 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.h:
```diff
@@ -0,0 +1,71 @@
+#include <wtf/Forward.h>
+#include <wtf/Noncopyable.h>
+#include <wtf/WeakPtr.h>
+namespace WebCore {
```
Blank line between includes and namespace

djg commented on Source/WebCore/Modules/identity/IdentityCredentialsContainer.cpp:
```diff
@@ -83,11 +84,14 @@ void IdentityCredentialsContainer::get(CredentialRequestOptions&& options, Crede
     }
 
-    // FIXME: Implement the actual get logic.
+    // if (options.mediation != MediationRequirement::Required) {
```
Remove commented code or add a FIXME note to explain why it's left here.

aestes commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp:
```diff
@@ -0,0 +1,108 @@
+
+void CredentialRequestCoordinator::discoverFromExternalSource(const Document& document, CredentialRequestOptions&& requestOptions, CredentialPromise&& promise)
+{
+
```
Nit: extra newline

aestes commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp:
```diff
@@ -0,0 +1,108 @@
+
+    if (requestOptions.signal) {
+        requestOptions.signal->addAlgorithm([weakThis = WeakPtr { *this }](JSC::JSValue) mutable {
+            if (!weakThis)
```
```suggestion
            RefPtr protectedThis = weakThis.get();
            if (!protectedThis)
```

(... then use `protectedThis` everywhere else you currently use `weakThis`)

aestes commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp:
```diff
@@ -0,0 +1,108 @@
+            if (!weakThis)
+                return;
+            weakThis->m_isCancelling = true;
+            weakThis->m_client->cancel([weakThis = WTFMove(weakThis)]() mutable {
```
Since it's a `unique_ptr`, are you sure `m_client` will be non-null at this point? I know you check it for null on line 54, but could it become null between when `addAlgorithm` is called and when this lambda executes?

aestes commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp:
```diff
@@ -0,0 +1,108 @@
+        });
+    }
+
+    auto callback = [weakThis = WeakPtr { *this }, promise = WTFMove(promise), abortSignal = WTFMove(requestOptions.signal)](/*DigitalCredentialResponseData&& data,*/ ExceptionData&& exception) mutable {
```
`weakThis` is currently unused in this lambda, so you could remove it from the capture list.

aestes commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp:
```diff
@@ -0,0 +1,108 @@
+                return;
+            weakThis->m_isCancelling = true;
+            weakThis->m_client->cancel([weakThis = WTFMove(weakThis)]() mutable {
+                if (!weakThis)
```
Ditto, re: creating a `protectedThis`

aestes commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp:
```diff
@@ -0,0 +1,108 @@
+
+    if (m_isCancelling) {
+        m_queuedRequest = [weakThis = WeakPtr { *this }, weakFrame = WeakPtr { *frame }, requestOptions = WTFMove(requestOptions), callback = WTFMove(callback)]() mutable {
+            if (!weakThis || !weakFrame)
```
Ditto, re: creating a `protectedThis`

aestes commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp:
```diff
@@ -0,0 +1,108 @@
+            RefPtr frame = weakFrame.get();
+            if (!frame)
+                return;
+            weakThis->m_client->requestDigitalCredential(*weakFrame, options, WTFMove(callback));
```
Could `m_client` be null here?

aestes commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.h:
```diff
@@ -0,0 +1,69 @@
+
+public:
+    WEBCORE_EXPORT explicit CredentialRequestCoordinator(std::unique_ptr<CredentialRequestCoordinatorClient>&&);
+    WEBCORE_EXPORT void setClient(std::unique_ptr<CredentialRequestCoordinatorClient>&&);
```
Do we need both a constructor that takes a `CredentialRequestCoordinatorClient` and `setClient`? Right now it seems like `setClient` is not called (unless I missed it). In the future will the client change during the lifetime of a `CredentialRequestCoordinator`?

marcoscaceres replied:
Yeah, I don't imagine it will change... and can always add this back if needed later.

aprotyas commented on Source/WebCore/Modules/identity/CredentialRequestCoordinatorClient.h:
```diff
@@ -0,0 +1,57 @@
+struct DigitalCredentialRequestOptions;
+class LocalFrame;
+
+using DCRequestCompletionHandler = CompletionHandler<void(WebCore::ExceptionData&&)>;
```
I think the abbreviation here, while saving a lot of letters, is less canonical than the full form, particularly since this is part of the broad `WebCore` namespace.

```suggestion
using DigitalCredentialRequestCompletionHandler = CompletionHandler<void(WebCore::ExceptionData&&)>;
```

https://webkit.org/code-style-guidelines/#names-full-words

aprotyas commented on Source/WebCore/Modules/identity/IdentityCredentialsContainer.cpp:
```diff
@@ -50,8 +51,20 @@ void IdentityCredentialsContainer::get(CredentialRequestOptions&& options, Crede
+    // if (options.mediation != MediationRequirement::Required) {
+    //     promise.reject(Exception { ExceptionCode::NotAllowedError, "Meditation is required."_s });
+    //     return;
+    // }
```
Let's make this a tracked FIXME instead and remove the commented code.

The bugzilla will have a space for the code we should commit based on the resolution of WICG/digital-credentials#147.

```suggestion
    // FIXME: <https://webkit.org/b/bug-number> waiting on conclusion to https://github.com/WICG/digital-credentials/issues/147
```

aprotyas commented on Source/WebCore/Modules/identity/IdentityCredentialsContainer.cpp:
```diff
@@ -73,21 +80,15 @@ void IdentityCredentialsContainer::get(CredentialRequestOptions&& options, Crede
     }
 
-    // FIXME: Implement the actual get logic.
+    // FIXME: <https://webkit.org/b/277322> midiation requirement,
```
Typo
```suggestion
    // FIXME: <https://webkit.org/b/277322> mediation requirement,
```

marcoscaceres committed:
Digital Credential API needs a CredentialRequestCoordinator and clients skeletons
https://bugs.webkit.org/show_bug.cgi?id=270778
rdar://124874018

Reviewed by Abrar Rahman Protyasha.

Implements CredentialRequestCoordinator and clients skeletons.

* LayoutTests/http/wpt/identity/identitycredentialscontainer-get-basics.https.html:
* LayoutTests/http/wpt/identity/identitycredentialscontainer-get-hidden.https.html:
* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/allow-attribute.https.html:
* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/enabled-on-self-origin-by-permissions-policy.https.sub.html:
* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/support/iframe.html:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp: Added.
(WebCore::CredentialRequestCoordinator::CredentialRequestCoordinator):
(WebCore::CredentialRequestCoordinator::discoverFromExternalSource):
* Source/WebCore/Modules/identity/CredentialRequestCoordinator.h: Copied from Source/WebCore/Modules/identity/IdentityCredentialsContainer.h.
* Source/WebCore/Modules/identity/CredentialRequestCoordinatorClient.h: Copied from Source/WebCore/Modules/identity/IdentityCredentialsContainer.h.
* Source/WebCore/Modules/identity/IdentityCredentialProtocol.h:
* Source/WebCore/Modules/identity/IdentityCredentialProtocol.idl:
* Source/WebCore/Modules/identity/IdentityCredentialsContainer.cpp:
(WebCore::IdentityCredentialsContainer::get):
* Source/WebCore/Modules/identity/IdentityCredentialsContainer.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Page.cpp:
* Source/WebCore/page/Page.h:
(WebCore::Page::credentialRequestCoordinator):
* Source/WebCore/page/PageConfiguration.cpp:
* Source/WebCore/page/PageConfiguration.h:

Canonical link: https://commits.webkit.org/281646@main

achristensen07 commented on Source/WebCore/page/Page.cpp:
```diff
@@ -398,6 +399,7 @@ Page::Page(PageConfiguration&& pageConfiguration)
 #endif
 #if ENABLE(WEB_AUTHN)
     , m_authenticatorCoordinator(makeUniqueRef<AuthenticatorCoordinator>(WTFMove(pageConfiguration.authenticatorCoordinatorClient)))
+    , m_credentialRequestCoordinator(makeUniqueRef<CredentialRequestCoordinator>(WTFMove(pageConfiguration.credentialRequestCoordinatorClient)))
```
This was the cause of the assertions.  CredentialRequestCoordinator is RefCounted, so we can't have a UniqueRef/unique_ptr to it managing its lifetime.

aprotyas replied:
Yes, there's actually no reason for it to be ref counted. It's only ever owned uniquely.
